lib/Controls.tsx

```diff
@@ -98,6 +191,159 @@ export const Controls: React.FC<ControlsProps> = ({
     [setVolume],
   )
 
+  const [totalDuration, setTotalDuration] = useState(duration)
+  const __mediaTimeline = useRef({
+    startDateTime:
+      startTime !== undefined ? DateTime.fromISO(startTime) : undefined,
+  })
+
+  /**
+   * Progress
+   *
+   * Compute progress of played and buffered amounts of media. This includes any
+   * media before the actual start of the video.
+   *
+   * The range on videoProperties specifies where we started to play (meaning,
+   * the time corresponding to currentTime = 0), and where the playback stops.
+   * To avoid having to collect extra data about the actual media length, we
+   * treat the end of the range as the end of the actual media (i.e. a simple
+   * way to establish the duration).
+   *
+   * Example:
+   *  - range = [0, undefined] => start from the beginning, unknown end
+   *  - range = [8, 19] => start from 8s into the media, stop at 19s in which
+   *    case currentTime = 0 is actually 8s. In this case the media is actually
+   *    25s long, but we cannot display that in our progress. So this system
+   *    only works correctly when playing back from any starting point till the
+   *    end of the media (i.e. no "chunks" within).
+   *
+   *    media        0 ------------------------------------------------- 25s
+   *    range                     8s ----------------------------- 19s
+   *    currentTime               0s ----------------------------- 11s
+   *    progress     0 ------------------------------------------- 19s
+   *
+   *  So we treat the start of the range as offset for total progress, and the
+   *  end of the range as total duration. That means we do not handle situations
+   *  where the duration is longer than the end of the range.
+   *
+   * When computing progress, if the duration is Infinity (live playback), we
+   * use the total buffered time as a (temporary) duration.
+   */
+  const [progress, setProgress] = useState({
+    playedFraction: 0,
+    bufferedFraction: 0,
+    counter: '',
+  })
+  useEffect(() => {
+    if (videoProperties === undefined) {
+      return
+    }
+    const { el, pipeline, range } = videoProperties
+    if (el === null || pipeline === undefined) {
+      return
+    }
+
+    // Extract range and update duration accordingly.
+    const [start = 0, end = duration] = range ?? [0, duration]
+    const __duration = duration ?? end ?? Infinity
+    setTotalDuration(__duration)
+
+    const updateProgress = () => {
+      const played = start + pipeline.currentTime
+      const buffered =
+        isHTMLMediaElement(el) && el.buffered.length > 0
+          ? start + el.buffered.end(el.buffered.length - 1)
+          : played
+      const total = __duration === Infinity ? buffered : __duration
```

Comment:
I suggest that you set `total` to `buffered` `if buffered > __duration`. Incase you are playing an ongoing recording the buffered will eventually go higher than the duration that was set when started playing.

Reply:
An ongoing recording should compute the total as the initial duration plus the elapsed time since you started playing, even when `buffered < __duration` (as the bar should increase already at the beginning of the recording). None of that was however included in this example (it doesn't work for ongoing recordings, only live caching).